src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,431 @@
+        {
+            Debug.Assert(byteCount >= 0);
+            Debug.Assert(limit >= 0);
+
+            if (ActiveMemory.Length >= limit)
+            {
+                // Already past limit. Do nothing.
+                return;
+            }
+
+            // Enforce the limit.
+            byteCount = Math.Min(byteCount, limit - ActiveMemory.Length);
+
+            EnsureAvailableSpace(byteCount);
+        }
+
+        public void EnsureAvailableSpace(int byteCount)
+        {
+            Debug.Assert(byteCount >= 0);
+
+            if (byteCount > AvailableMemory.Length)
+            {
+                GrowAvailableSpace(byteCount);
+            }
+        }
+
+        public void GrowAvailableSpace(int byteCount)
+        {
+            Debug.Assert(byteCount > AvailableMemory.Length);
+
+            CheckState();
+
+            uint ubyteCount = (uint)byteCount;
+
+            uint newBytesNeeded = ubyteCount - (uint)AvailableMemory.Length;
+            uint newBlocksNeeded = (newBytesNeeded + BlockSize - 1) / BlockSize;
+
+            // Ensure we have enough space in the block array for the new blocks needed.
+            if (_blocks is null)
+            {
+                Debug.Assert(_allocatedEnd == 0);
+                Debug.Assert(_activeStart == 0);
+                Debug.Assert(_availableStart == 0);
+
+                int blockArraySize = 4;
+                while (blockArraySize < newBlocksNeeded)
+                {
+                    blockArraySize *= 2;
+                }
+
+                _blocks = new byte[]?[blockArraySize];
+            }
+            else
+            {
+                Debug.Assert(_allocatedEnd % BlockSize == 0);
+                Debug.Assert(_allocatedEnd <= _blocks.Length * BlockSize);
+
+                uint allocatedBlocks = _allocatedEnd / BlockSize;
+                uint blockArraySize = (uint)_blocks.Length;
+                if (allocatedBlocks + newBlocksNeeded > blockArraySize)
+                {
+                    // Not enough room in current block array.
+                    uint unusedInitialBlocks = _activeStart / BlockSize;
+                    uint usedBlocks = (allocatedBlocks - unusedInitialBlocks);
+                    uint blocksNeeded = usedBlocks + newBlocksNeeded;
+                    if (blocksNeeded > blockArraySize)
+                    {
+                        // Need to allocate a new array and copy.
+                        while (blockArraySize < blocksNeeded)
+                        {
+                            blockArraySize *= 2;
+                        }
+
+                        byte[]?[] newBlockArray = new byte[]?[blockArraySize];
+                        _blocks.AsSpan().Slice((int)unusedInitialBlocks, (int)usedBlocks).CopyTo(newBlockArray);
+                        _blocks = newBlockArray;
+                    }
+                    else
+                    {
+                        // We can shift the array down to make enough space
```

Comment:
Did you consider making the byte[][] circular rather than shifting like this?

Reply:
Yes, that seems like a potentially good optimization. I elected not to do it for now.